src/bio/bio_wal.c

```diff
@@ -979,27 +979,628 @@ flush_wal_header(struct bio_meta_context *mc)
+	bio_sgl_fini(&bsgl);
+
+	return rc;
+}
+
+static int
+verify_tx_hdr(struct wal_super_info *si, struct wal_trans_head *hdr, uint64_t tx_id)
+{
+	bool	committed = false;
+
+	if (wal_id_cmp(si, tx_id, si->si_commit_id) <= 0)
+		committed = true;
+
+	if (hdr->th_magic != WAL_HDR_MAGIC) {
+		D_CDEBUG(committed, DLOG_ERR, DB_IO, "Mismatched WAL head magic, %x != %x\n",
+			 hdr->th_magic, WAL_HDR_MAGIC);
+		return committed ? -DER_INVAL : 1;
+	}
+
+	if (hdr->th_gen != si->si_header.wh_gen) {
+		D_CDEBUG(committed, DLOG_ERR, DB_IO, "Mismatched WAL generation, %u != %u\n",
+			 hdr->th_gen, si->si_header.wh_gen);
+		return committed ? -DER_INVAL : 1;
+	}
+
+	if (id2seq(hdr->th_id) < id2seq(tx_id)) {
+		D_CDEBUG(committed, DLOG_ERR, DB_IO, "Stale sequence number detected, %u < %u\n",
+			 id2seq(hdr->th_id), id2seq(tx_id));
+		return committed ? -DER_INVAL : 1;
+	} else if (id2seq(hdr->th_id) > id2seq(tx_id)) {
+		D_ERROR("Invalid sequence number detected, %u > %u\n",
+			id2seq(hdr->th_id), id2seq(tx_id));
+		return -DER_INVAL;
+	}
+
+	if (hdr->th_id != tx_id) {
+		D_ERROR("Mismatched transaction ID. "DF_U64" != "DF_U64"\n", hdr->th_id, tx_id);
+		return -DER_INVAL;
+	}
+
+	if (hdr->th_tot_ents == 0) {
+		D_ERROR("Invalid entry number\n");
+		return -DER_INVAL;
+	}
+
+	return 0;
+}
+
+static int
+verify_data(struct bio_meta_context *mc, uint64_t off, uint32_t len, uint32_t expected_csum,
+	    char **dbuf, unsigned int *dbuf_len)
+{
+	struct bio_sglist	 bsgl;
+	struct bio_iov		*biov;
+	d_sg_list_t		 sgl;
+	d_iov_t			 iov;
+	char			*buf;
+	unsigned int		 iov_sz = (1UL << 20);	/* 1MB */
+	unsigned int		 read_sz, tot_read = len;
+	bio_addr_t		 addr = { 0 };
+	uint32_t		 csum;
+	int			 iov_nr, csum_len, rc;
+
+	D_ASSERT(len > 0);
+	if (*dbuf_len < len) {
+		D_FREE(*dbuf);
+		*dbuf = NULL;
+		*dbuf_len = 0;
+
+		D_ALLOC(buf, max(iov_sz, len));
+		if (buf == NULL)
+			return -DER_NOMEM;
+		*dbuf = buf;
+		*dbuf_len = max(iov_sz, len);
+	} else {
+		buf = *dbuf;
+	}
+
+	d_iov_set(&iov, buf, len);
+	sgl.sg_iovs = &iov;
+	sgl.sg_nr = 1;
+	sgl.sg_nr_out = 0;
+
+	/* Read in 1MB sized IOVs */
+	iov_nr = (len + iov_sz - 1) / iov_sz;
+	rc = bio_sgl_init(&bsgl, iov_nr);
+	if (rc)
+		return rc;
+
+	while (tot_read > 0) {
+		biov = &bsgl.bs_iovs[bsgl.bs_nr_out];
+
+		bio_addr_set(&addr, DAOS_MEDIA_NVME, off);
+		read_sz = min(tot_read, iov_sz);
+		bio_iov_set(biov, addr, read_sz);
+
+		bsgl.bs_nr_out++;
+		tot_read -= read_sz;
+		off += read_sz;
+		D_ASSERT(bsgl.bs_nr_out <= iov_nr);
+	}
+
+	rc = bio_readv(mc->mc_data, &bsgl, &sgl);
+	bio_sgl_fini(&bsgl);
+	if (rc) {
+		D_ERROR("Read data from data blob failed. "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	csum_len = meta_csum_len(mc);
+	rc = meta_csum_calc(mc, buf, len, &csum, csum_len);
+	if (rc) {
+		D_ERROR("Calculate data csum failed. "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	if (csum != expected_csum) {
+		D_DEBUG(DB_IO, "Mismatched data csum, %u != %u\n", csum, expected_csum);
+		return 1;
```

Comment:
-DER_MISMATCH?

Reply:
It means the data isn't committed (so this transaction isn't committed), and we return 1 to stop replay.